status/status.go

```diff
@@ -128,3 +130,39 @@ func FromError(err error) (s *Status, ok bool) {
 	}
 	return nil, false
 }
+
+// WithDetails returns a new status with the provided details messages appended to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.Proto() != nil.
+	p := s.Proto()
+	for _, detail := range details {
+		any, err := ptypes.MarshalAny(detail)
+		if err != nil {
+			return nil, err
+		}
+		p.Details = append(p.Details, any)
+	}
+	return &Status{s: p}, nil
+}
+
+// Details returns a slice of details messages attached to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) Details() ([]proto.Message, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.s != nil.
+	details := make([]proto.Message, 0, len(s.s.Details))
+	for _, any := range s.s.Details {
+		detail := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+			return nil, err
+		}
```

Comment:
What about 

4. return []proto.Message, but also, either
   4a. return a second []error slice; which is non-nil exactly where the first slice is nil; or
   4b. add a DetailsErrors method that returns []error.

I kind of like 4b because preserves type-safety, handles the common case (people won't care about the error usually), but makes it easy to get the errors if needed (without duplicating the body of Details).